### ISSUE TITLE
After revoking a credential at the issuer side, the verifier can still successfully verify a credential sent from the holder

### ISSUE THREAD

**Issue by yunxi-zhang** [hyperledger/aries-cloudagent-python#1020]
I'm running two Acapy agents (v0.6), one as an issuer/verifier and the other as the holder on a von dev network + a tails file server. Once the issuer issued a credential and gave it to the holder and holder shared the presentation back with the issuer as a verifier, the whole credential verification process works following the 3 steps works: 1) verifier sends a presentation request 2) holder shares the presentation and 3)verifier successfully verifies the credential. 

However, after revoking this credential by calling "/revocation/revoke" api and checking the credential's status has been updated to "revoked" by calling the "revocation/credential-record" api on the issuer, when I rerun the 3-steps credential verification process again, the verification result still shows the revoked credential passed credential verification.

The expected result should be: the revoked credential should **NOT** pass the credential verification.

**Comment by swcurran** [comment]
I'm not sure of how the mechanism works but I believe ACA-Py supports a "deferred revocation publishing" model, where you can revoke one more credentials and then using another call, publish all the revocations in a batch.  I think there is a way to do "revoke and publish", but am not sure of the flags on the endpoint.

Please check that.  @esune may be able to provide additional guidance if that is not the issue.

**Comment by yunxi-zhang** [comment]
Thanks @swcurran .

Hello @esune , please let me know if you need more info. Just add one more detail here, when I call the api  "/revocation/revoke", I set the flag for "publish" as true in the JSON body, so assuming this would auto publish the revoked credential to the ledger. Please let me know if I miss any necessary other steps here for credential revocation.

**Comment by esune** [comment]
@yunxi-zhang can you please provide a sample of the proof-request you are sending to the holder for me to look at?

**Comment by esune** [comment]
@yunxi-zhang not exactly. What you are showing me is what the `holder` will compile and send to the `verifier`. What I would like to know is what the `verifier` is asking the holder to present, usually this is prepared in the `/present-proof/create-request` or the `/present-proof/send-request` endpoints (I am guessing the latter in your case, as you must be using an existing connection).

**Comment by yunxi-zhang** [comment]
Apologies, misread your message. Below is the JSON payload sent by the verifier: 

```
{
    "comment": "string",
    "connection_id": "f433f333-f2d2-45a1-951d-059e30bbd2c6",
    "proof_request": {
        "name": "Proof request",
        "nonce": "1234567890",
        "requested_attributes": {
            "additionalProp1": {
                "name": "score"
            }
        },
        "requested_predicates": {
          
        },
        "version": "1.0"
    },
    "trace": false
}
```

**Comment by yunxi-zhang** [comment]
Thanks for the info @esune. Still bit confused with the semantics of the `non-revocation` attribute. If I specify the epoch for `from` and `to`, does it mean during the time period between `from` to `to`, revocation check for the credential won't happen?

**Comment by swcurran** [comment]
Take a look at this -- https://github.com/hyperledger/indy-hipe/tree/master/text/0011-cred-revocation#proof-of-non-revocation-intervals

It is confusing, so you are not alone.  You can also look at this:

https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0441-present-proof-best-practices

**Comment by yunxi-zhang** [comment]
Thanks for the two useful links @swcurran. @esune, just wondering if "now()" can be used as a proper value for the  `non-revocation` attribute when calling the api like below?

```
{
    "comment": "string",
    "connection_id": "f433f333-f2d2-45a1-951d-059e30bbd2c6",
    "proof_request": {
        "name": "Proof request",
        "nonce": "1234567890",
        "requested_attributes": {
            "additionalProp1": {
                "name": "score"
            }
        },
        "requested_predicates": {},
        "non_revoked": {
            "from": "**now()**",
            "to":"**now()**"
        },
        "version": "1.0"
    },
    "trace": false
}
```

**Comment by swcurran** [comment]
You need the epoch time returned from a current time function.  A number.

**Comment by yunxi-zhang** [comment]
Just tested it, it has to be a number, lol. Thanks @swcurran for the docs links, very useful. @esune, just tested by putting the `non-revocation` attribute with the current time, it works now.

**Comment by esune** [comment]
> 
> 
> Thanks for the info @esune. Still bit confused with the semantics of the `non-revocation` attribute. Q1: If I specify the epoch for `from` and `to`, does it mean during the time period between `from` to `to`, revocation check for the credential won't happen, is this correct? Q2: without having this `non-revocation` attribute, revocation check for the credential will never happen, is this correct?

A1: This is what I meant when I said `omit from`. The interval is not really an interval, I am unable to explain why it is structured that way. For the purpose of verifying the revocation status, just worry about setting the value of `to` in the proof-request. This will result on the verifier checking that the state of the credential being presented was "not revoked" at the time identified by the epoch timestamp that `to` was set to.
A2: Correct. Omitting a non-revocation interval from the proof-request results in the revocation state not being checked at all at verification time.

**Comment by yunxi-zhang** [comment]
@esune, when you said "The interval is not really an interval" confused me again. After reading the docs, my understanding on the `from` and `to` are used for the verifier to specify whether the credential is still non-revoked for the time interval, is this correct? I see there are differences between the semantics of `from` and `to` between the Aries and Indy node, but as a acapy user, I don't need to worry on this difference, right?  

I may still not fully understand the concept here, and I've got another question now. For instance, if an issuer issued a credential at time a, and revoked it at time b. When a verifier puts the from as 0, which means "1 January 1970 00:00:00" and then put a time c as the value for to. What will be the credential verification result? 

Time line:
Time 0 ------->Time a------------------>Time b------------------->Time c
      |                        |                                             |                                           |
   from                credential issued            credential revoked                      to

**Comment by swcurran** [comment]
`from` and `to` were _intended_ to be used for an interval where the verifier is saying "prove to me the credential was not revoked at some point in this period".  However, in practice, that is not possible. That's why the "best practices" RFC 441 was put in place, to say, "intervals were a good idea, but the don't work, so always use 0 for from" and put the time you want in "to", and the holder/prover should just use to as the "point in time" (not interval) to demonstrate non-revocation.  Further, since most use cases want to know if  the credential is revoked now, they use "to" as `now()`.

Does that make sense?

**Comment by yunxi-zhang** [comment]
@swcurran, do you mean the real credential revocation check will only be done against the two time points -`from` and `to` only? if that's the case, in my previous example, what will be the result for a credential verification, if I put time a for `from` and time b for `to`?

**Comment by yunxi-zhang** [comment]
Thanks for the clarification. So, basically, only the time point put in `to` is actually used in acapy for the credential revocation checking in practice for now, is this correct? 

Regarding your words "the **holder/prover** should just use to as the "point in time" (not interval)", the suggestion given by @esune for resolving my issue is that the **verifier** should set the `to` value in its send-request, is here a conflict?

**Comment by sklump** [comment]
Historical context: the original non-revocation period was a courtesy for provers: as long as they had a revocation registry state calculated anywhere in the interval, they could use it to build proof; otherwise, they would need to compute one anew.
Comments describe an old cache implementation along these lines:
https://github.com/PSPC-SPAC-buyandsell/von_anchor/blob/9d3f96451c1c7a2fee8c0141fc5f32f3596791d0/von_anchor/cache.py#L442

**Comment by swcurran** [comment]
> Regarding your words "the **holder/prover** should just use to as the "point in time" (not interval)", the suggestion given by @esune for resolving my issue is that the **verifier** should set the `to` value in its send-request, is here a conflict?

Listen to @esune -- he knows the code way better than I!  I'm more a "concepts" person :-).

**Comment by esune** [comment]
> 
> 
> > Regarding your words "the **holder/prover** should just use to as the "point in time" (not interval)", the suggestion given by @esune for resolving my issue is that the **verifier** should set the `to` value in its send-request, is here a conflict?
> 
> Listen to @esune -- he knows the code way better than I! I'm more a "concepts" person :-).

I just got burnt by this so many times I try to save the pain to others, however I realize that I may be making assumptions when explaining things 🙂 

The reason to ignore setting `from` is based on the fact that, as discussed above, while the non-revocation time frame was meant to be an interval, it doesn't exactly work that way in practice.

In the first link I sent ([this](https://github.com/hyperledger/aries-cloudagent-python/blob/main/docs/GettingStartedAriesDev/CredentialRevocation.md)) in the example (scroll to bottom) it states that `"from": <seconds from Unix Epoch> # Optional, default is 0`. This is why my recommendation is - in general - to ignore setting `from` and only worry about `to`. This will translate to the verifier checking whether `the credential was revoked at any time before time ttttttt`, where time `ttttttt` is the epoch timestamp that was set as value for `to` in the first place.

----

**_Deep dive on revocation intervals_**

Seeing the non-revocation interval without knowing the underlying logic (as I did as well), you would think that specifying a non-revocation interval means that the credential being verified must have been non-revoked in that lapse of time. However, based on the Indy implementation, the non-revocation interval should be rather read as `there is no revocation information published by the issuer in the specified lapse of time, for the specified credential`. This is where the "fun" begins.

**Example:** Credential `C1` is issued at time `T0`. The issuer revokes the credential at time `T1` (and publishes revocation information to the ledger at this time as well) and a verifier wants to verify that the credential is valid during interval `T2-T3`, where `T0 < T1 < T2 <=T3`. The proof request would read something like `from: T2, to: T3` and the assumption would be that in the lapse of time between `T2` and `T3` the credential must have been non-revoked. The problem arises when verification kicks in as the last published revocation timestamp `T1` is outside of the interval `T2-T3` and therefore the credential will be considered non-revoked (I do not know if this changed over the past few months, I don't think so).

So, based on the above example, since it is impossible (or at least not trivial and not computational expensive) to know when the last revocation was published by an issuer to the ledger, or at least when was the first revocation published by the issuer to the ledger, completely omitting `from` and considering the interval as a point in time specified by `to` is what will be the most common use-case. And based on the above example, it is NOT possible to check whether a credential was valid for an entire, well-defined time frame as mismatches between the time frame and the last published revocation will lead to false negatives.
